design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+    image: rust:latest
+    steps:
+      - add_musl: rustup target install x86_64-unknown-linux-musl
+      - build: cargo build --release --target=x86_64-unknown-linux-musl
+```
+
+### Design considerations
+
+1. Resource usage
+   1. Copying to and from a remote server for every job in a pipeline can lead to unnecessary network bandwidth usage and unnecessarily slow speeds.
+2. Integrity and consistency of cached data.
+3. User experience
+   1. Is it easily adaptable to new use cases?
+   2. Is it easily enabled and administered?
+   3. Is it easily configurable by end users once enabled?
+   4. Does it speed up build time? How much does it speed up build time for various use cases?
+   5. Is it consistent?
+   6. Is it reliable?
+4. Security
+   1. Is it possible to inject malicious code into pipelines via the cache mechanism? Malicious pull requests are a good example of this.
+
+### Security
+
+Injecting malicious code into a build via the caching mechanism is a real concern. This can lead to theft of tokens or credentials from job, or in even lead to malicious code being deployed to a production environment. 
+
+Pull requests, as the main vector for injection of malicious code, require additional security considerations. Pull requests therefore treat caches as readonly. 
+
+Caching of potentially sensitive data also needs to be handled. Data provided via the secrets field should not be cached. 
+
+### Cache invalidation
+
+- [Proposed] Changing environment variables or secrets should invalidate existing caches. Mass invalidation can be avoided by including a hash of variables and secrets with the cache metadata.
+- [Proposed] The user should also be able to invalidate caches via the UI.
+
+### Conflicts
+
+Multiple caches trying to restore the same file/files will result in an error. Caches by default do not overwrite existing files or directories. This behaviour can be configured by setting selecting a merging strategy.
+
+The following merging strategies are defined for dealing with file conflicts when extracting cached data on ta local build node.
+
+1. merge: Compares creation time of files and directories. Recursively merges files and directories from multiple caches. Selects the *newest* version of a file.
+2. [Proposed] preserve: Compares creation time of files and directories. Recursively merges files and directories from multiple caches. Selects the *oldest* version of a file.
+3. reject: Fails the build if there are any file conflicts.
+
+
+### Scoping
+
+Cache scoping levels can be expressed as `Job level < Event level < Pipeline level`. Job level scoping limits the cache to particular jobs, within a single event, within a single pipeline. Event level scoping limits the cache to a certain event within a single pipeline. Pipeline level scoping represents a globally scoped cache for the single pipeline, with the cache being available for all instances of the pipeline.
+
+#### [Tentative] Regarding commits, branches and scoping
```

Comment:
Any comments on this? It's only tentative, so any indication of inclination would be appreciated. (Keep/disgard etc.)

Reply:
I think these could be valid use cases. Maybe we add more scoping in addition to event-level and pipeline-level (and implement them later)?